src/main/java/jp/co/soramitsu/sora/common/ArrayTree.java

```diff
@@ -0,0 +1,96 @@
+package jp.co.soramitsu.sora.common;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.Getter;
+
+public class ArrayTree<E> {
+
+  @Getter
+  private List<E> tree;
+
+  public ArrayTree(List<E> tree) {
+    this.tree = tree;
+  }
+
+  public boolean hasElement(int i) {
+    return i < tree.size() && get(i) != null;
+  }
+
+  public List<E> getLeafs() {
+    int leafsTotal = (tree.size() + 1) / 2;
+    int leftmostLeafIndex = leafsTotal - 1;
+    return tree.stream().skip(leftmostLeafIndex).collect(Collectors.toList());
+  }
+
+  public int find(E el) {
+    for (int pos = 0; pos < tree.size(); pos++) {
+      if (tree.get(pos).equals(el)) {
+        // found
+        return pos;
+      }
+    }
+
+    return -1;
+  }
+
+  public int size() {
+    return tree.size();
+  }
+
+  public E get(int i) {
+    return tree.get(i);
+  }
+
+  public void set(int i, E el) {
+    tree.set(i, el);
```

Comment:
validate that i is within tree size, throw otherwise

Reply:
underlying List does this check

Reply:
List interface does not enforce such check on implementation